### PR TITLE
Feeling my way around a HA NN setup and populated dfs.nameservices but no…

### PR DESCRIPTION
**Pull request by acqant**
…t the rest

default['hadoop']['hdfs_site']['dfs.nameservices']='kitchenCluster'
default['hadoop']['hdfs_site']['dfs.ha.namenodes']='kitchenCluster,kitchenCluster_TW0'

This PR checks the intention of doing HA by seeing dfs.ha.namenodes configured but not dfs.ha.namenodes.#{ns}.  Not sure if the unless should be removed or now.

default['hadoop']['hdfs_site']['dfs.ha.namenodes.kitchenCluster']='10.0.0.2,10.0.0.3'
default['hadoop']['hdfs_site']['dfs.namenode.rpc-address.kitchenCluster.10.0.0.2']='10.0.0.2:8020'
default['hadoop']['hdfs_site']['dfs.namenode.rpc-address.kitchenCluster.10.0.0.3']='10.0.0.3:8020'
default['hadoop']['hdfs_site']['dfs.namenode.http-address.kitchenCluster.10.0.0.2']='10.0.0.2:50070'
default['hadoop']['hdfs_site']['dfs.namenode.http-address.kitchenCluster.10.0.0.3']='10.0.0.3:50070'

### CODE REVIEW

File: recipes/_hadoop_hdfs_ha_checkconfig.rb [hunk 1]
```diff
@@ -27,6 +27,9 @@
 # We have dfs.nameservices, and now need to check them
 dfs_nameservices.each do |ns|
   # Start namenode checks
+  if node['hadoop']['hdfs_site']['dfs.ha.namenodes'] 
+    Chef::Application.fatal!("NameNode HA requires node['hadoop']['hdfs_site']['dfs.ha.namenodes.#{ns}']") unless node['hadoop']['hdfs_site'].key?("dfs\.ha\.namenodes\.#{ns}")
+  end
```
Comment (wolf31o2) [inline recipes/_hadoop_hdfs_ha_checkconfig.rb]: These aren't checking the same thing. What are you actually trying to do here?

Comment (acqant) [inline recipes/_hadoop_hdfs_ha_checkconfig.rb]: Following this trying to setup a HA name node with journal nodes
https://hadoop.apache.org/docs/r2.4.1/hadoop-project-dist/hadoop-hdfs/HDFSHighAvailabilityWithQJM.html

Using the cookbook it converged but I noticed the hdfs-site config didn't
have the settings below even though I intended to do HA by setting
hdfs.ha.namenodes and nothing else.

*dfs.nameservices* - the logical name for this new nameservice
*dfs.ha.namenodes.[nameservice ID]* - unique identifiers for each NameNode
in the nameservice
​*dfs.namenode.rpc-address.[nameservice ID].[name node ID]* - the
fully-qualified RPC address for each NameNode to listen on
*dfs.namenode.http-address.[nameservice ID].[name node ID]* - the
fully-qualified HTTP address for each NameNode to listen on​
​*dfs.namenode.shared.edits.dir* - the URI which identifies the group of
JNs where the NameNodes will write/read edits
*dfs.client.failover.proxy.provider.[nameservice ID]* - the Java class that
HDFS clients use to contact the Active NameNode
*dfs.ha.fencing.methods* - a list of scripts or Java classes which will be
used to fence the Active NameNode during a failover
*fs.defaultFS* - the default path prefix used by the Hadoop FS client when
none is given
*dfs.journalnode.edits.dir* - the path where the JournalNode daemon will
store its local state​

Comment (acqant) [inline recipes/_hadoop_hdfs_ha_checkconfig.rb]: Put another way it says, "Hey you set  dfs.nameservers and you set
dfs.ha.namenodes but don't have dfs.ha.namenodes.NameServer"

It silently failed/continued on me.

Comment (acqant) [inline recipes/_hadoop_hdfs_ha_checkconfig.rb]: ​dfs.nameservices that should be.​

Comment (wolf31o2) [inline recipes/_hadoop_hdfs_ha_checkconfig.rb]: That was really confusing to read. Anyway, there's no such property `dfs.ha.namenodes` in `hdfs-site.xml` which means your `if` statement should never be matched. It makes your entire code block pointless.

https://hadoop.apache.org/docs/r2.4.1/hadoop-project-dist/hadoop-hdfs/hdfs-default.xml

It sounds like you're attempting to setup HA without actually configuring all of the required options, but I really cannot tell.

Comment (acqant) [inline recipes/_hadoop_hdfs_ha_checkconfig.rb]: Probably, do you see a logic issue though?

I've intended to do ha by fulfilling this
if node['hadoop'].key?('hdfs_site') &&
node['hadoop']['hdfs_site'].key?('dfs.nameservices')


but the required settings are getting skipped because of this line
next unless node['hadoop']['hdfs_site'].key?("dfs\.ha\.namenodes\.#{ns}")

You need to set
dfs.ha.namenodes.nameservice
which I did not and it converged without error and without the required
values.

**Comment by wolf31o2** [comment]
Here's the problem... `dfs.nameservices` does *not* mean "NameNode HA"... it means "multiple name services" and is usable for HA and/or federation... only HA requires settings under `dfs.ha.namenodes.#{ns}` as federated namenode does not. Your changes will break federated namenode, so your logic is bad.

It needs to be something like:
is `dfs.nameservices` configured?
for each item under `dfs.nameservices` is it for HA or federation?
if HA, do we have all HA configs
if federated, do we have all federated configs (currently not checked, at all)

This means it is 100% valid to have `dfs.nameservices` configured **without** `dfs.ha.namenodes.#{ns}` on a cluster... just not an HA cluster.

Also, you do not seem to understand how to setup HA NameNode using this cookbook. It **requires** the user to provide all of the correct attributes. Nothing is done for you in this cookbook. This recipe is nothing more than a sanity check on your configuration.

**Comment by wolf31o2** [comment]
If you look at the test output, you'll see that the logic is broken. You're triggering a failure for any setup that is not HA.

**Comment by acqant** [comment]
ok, I hear you.  Just close that PR then.  I think I know how to get this
cookbook to do what I need at this point but I still have a few open
questions that trial and error will sort out.

Why can't you submit issues?  So many questions can be answered by
submitting issues and updating the Readme that would make your life and
other end users life better.

**Comment by wolf31o2** [comment]
You can submit issues... We don't use GitHub issues because they're nearly impossible to plan cross-project. There was a link to [guidelines for contributing](CONTRIBUTING.md) when you created your pull request which is common practice in GitHub repositories.

**Comment by wolf31o2** [comment]
Meh... that relative link is broken... it's here

https://github.com/caskdata/hadoop_cookbook/blob/master/CONTRIBUTING.md

**Comment by acqant** [comment]
oh great!
So I'm using test-kitchen vagrant and private_network to test a 3 node
zookeeper/journalnode, 2 node namenode and 1 node datanode wrapper
cookbook.  Looks like it's all working except for the zookeeper myid logic.
  It's a total edge case where dns doesn't work and you're using a second
non-default nic, not sure what options are left other than hack /etc/hosts
or the zookeeper_server.rb.  Any suggestions?

**Comment by wolf31o2** [comment]
Your only real option is to setup `/etc/hosts` to match what Hadoop will expect. You'll need this for the Hadoop services to function, anyway. You **can** hack Hadoop via configuration to kinda ignore forward/reverse lookups, but it's really a pain and not a good tested path, so I wouldn't use it, personally.

**Comment by wolf31o2** [comment]
Closing this... I filed a JIRA with the gist of this PR and the problem so that I will remember to make the HA configuration checking a bit more robust.

https://issues.cask.co/browse/COOK-61
